src/middleware.rs

```diff
@@ -58,74 +55,49 @@ impl PluginFor<Response> for HandlebarsEngine {
-        info!("Failed to open file {}, skipped.", path.display());
-        None
+
+impl HandlebarsEngine {
+    /// #[Deprecated], for backward compaitibility only
+    pub fn new(prefix: &str, suffix: &str) -> HandlebarsEngine {
+        let mut hbs = HandlebarsEngine::new2();
+        hbs.add(Box::new(DirectorySource::new(prefix, suffix)));
+        hbs.reload();
+        hbs
     }
-}
 
-fn filter_file(entry: &DirEntry, suffix: &str) -> bool {
-    entry.file_name().to_str()
-        .and_then(|s| Some(s.starts_with(".") || s.starts_with("#") || !s.ends_with(suffix)))
-        .unwrap_or(false)
-}
+    pub fn from(prefix: &str, suffix: &str, custom: Handlebars) -> HandlebarsEngine {
+        let mut hbs = HandlebarsEngine::of(custom);
+        hbs.add(Box::new(DirectorySource::new(prefix, suffix)));
+        hbs.reload();
+        hbs
+    }
 
-impl HandlebarsEngine {
-    pub fn reload(&self) {
-        let mut hbs = self.registry.write().unwrap();
-        match current_dir()  {
-            Ok(current_path) => {
-                let mut prefix_path = PathBuf::from(current_path);
-                prefix_path.push(self.prefix.clone());
-                let template_path = prefix_path.as_path();
-
-                info!("Loading templates from path {}", template_path.display());
-                let walker = WalkDir::new(template_path);
-
-                hbs.clear_templates();
-                let suffix = &self.suffix;
-                let prefix_len = template_path.as_os_str().to_str().unwrap().len();
-                for p in walker.min_depth(1).into_iter().filter(|e| e.is_ok() && !filter_file(e.as_ref().unwrap(), suffix)).map(|e| e.unwrap()) {
-                    let path = p.path();
-                    let disp = path.to_str().unwrap();
-                    debug!("getting file {}", disp);
-                    let tpl_name = &disp[prefix_len .. disp.len()-suffix.len()];
-                    if let Some(tpl) = read_file(&path) {
-                        if let Err(e) = hbs.register_template_string(tpl_name, tpl){
-                            warn!("Failed to parse template {}, {}", tpl_name, e);
-                        } else {
-                            info!("Added template {}", tpl_name);
-                        }
-                    }
-                }
-            },
-            Err(e) => {
-                error!("Failed to get current directory due to {}", e);
-            }
+    pub fn new2() -> HandlebarsEngine {
+        HandlebarsEngine {
+            sources: Vec::new(),
+            registry: RwLock::new(Box::new(Handlebars::new()))
         }
     }
 
-    pub fn from(prefix: &str, suffix: &str, custom: Handlebars) -> HandlebarsEngine {
-        let eng = HandlebarsEngine {
-            prefix: prefix.to_string(),
-            suffix: suffix.to_string(),
-            registry: RwLock::new(Box::new(custom))
-        };
-        eng.reload();
-        eng
+    pub fn of(reg: Handlebars) -> HandlebarsEngine {
```

Comment:
This is precisely what is needed for full control. You have one API for 90% usecase, and the rest has to be done with `of`, why not?